File: orchestrator/safety.go
package main

import "strings"

// SafetyGate controls which domains are allowed and which need confirmation.
type SafetyGate struct {
	allowed               map[string]bool
	blocked               map[string]bool
	confirmOnly           map[string]bool
	confirmEntityKeywords []string
}

// NewSafetyGate creates a new safety gate with default policies.
func NewSafetyGate() *SafetyGate {
	return &SafetyGate{
		allowed: map[string]bool{
			"light":         true,
			"switch":        true,
			"cover":         true,
			"climate":       true,
			"media_player":  true,
			"fan":           true,
			"scene":         true,
			"script":        true,
			"automation":    true,
			"input_boolean": true,
			"input_select":  true,
			"input_number":  true,
			"input_text":    true,
			"vacuum":        true,
			"humidifier":    true,
			"water_heater":  true,
			"siren":         true,
			"remote":        true,
			"button":        true,
			"number":        true,
			"select":        true,
		},
		blocked: map[string]bool{
			"shell_command":           true,
			"rest_command":            true,
			"notify":                  true,
			"persistent_notification": true,
			"homeassistant":           true,
		},
		confirmOnly: map[string]bool{
			"lock":                true,
			"alarm_control_panel": true,
			"valve":               true,
		},
		confirmEntityKeywords: []string{
			"garage",
		},
	}
}

// IsAllowed returns true if the domain can be acted on.
func (sg *SafetyGate) IsAllowed(domain string) bool {
	if sg.blocked[domain] {
		return false
	}
	return sg.allowed[domain]
}

// NeedsConfirmation returns true if the domain requires explicit confirmation.
func (sg *SafetyGate) NeedsConfirmation(domain string) bool {
	return sg.confirmOnly[domain]
}

// NeedsEntityConfirmation enforces confirmation for risky targets identified by name/entity_id.
func (sg *SafetyGate) NeedsEntityConfirmation(entityID string) bool {
	lower := strings.ToLower(entityID)
	for _, keyword := range sg.confirmEntityKeywords {
		if strings.Contains(lower, keyword) {
			return true
		}
	}
	return false
}

File: orchestrator-rs/src/safety.rs
use crate::config::SafetyConfig;
use std::collections::HashSet;

pub struct SafetyGate {
    allowed: HashSet<String>,
    blocked: HashSet<String>,
    confirm_only: HashSet<String>,
    confirm_entity_keywords: Vec<String>,
}

impl Default for SafetyGate {
    fn default() -> Self {
        Self::new()
    }
}

impl SafetyGate {
    pub fn with_config(cfg: &SafetyConfig) -> Self {
        Self {
            allowed: cfg.allowed_domains.iter().cloned().collect(),
            blocked: cfg.blocked_domains.iter().cloned().collect(),
            confirm_only: cfg.confirm_domains.iter().cloned().collect(),
            confirm_entity_keywords: cfg.confirm_entity_keywords.clone(),
        }
    }

    pub fn new() -> Self {
        let allowed = vec![
            "light", "switch", "cover", "climate", "media_player", "fan", "scene", "script",
            "automation", "input_boolean", "input_select", "input_number", "input_text",
            "vacuum", "humidifier", "water_heater", "siren", "remote", "button", "number", "select",
        ]
        .into_iter()
        .map(String::from)
        .collect();

        let blocked = vec![
            "shell_command",
            "rest_command",
            "notify",
            "persistent_notification",
            "homeassistant",
        ]
        .into_iter()
        .map(String::from)
        .collect();

        let confirm_only = vec!["lock", "alarm_control_panel", "valve"]
            .into_iter()
            .map(String::from)
            .collect();

        let confirm_entity_keywords = vec!["garage".to_string()];

        Self {
            allowed,
            blocked,
            confirm_only,
            confirm_entity_keywords,
        }
    }

    pub fn is_allowed(&self, domain: &str) -> bool {
        if self.blocked.contains(domain) {
            return false;
        }
        self.allowed.contains(domain)
    }

    pub fn needs_confirmation(&self, domain: &str) -> bool {
        self.confirm_only.contains(domain)
    }

    pub fn needs_entity_confirmation(&self, entity_id: &str) -> bool {
        let lower = entity_id.to_lowercase();
        self.confirm_entity_keywords
            .iter()
            .any(|kw| lower.contains(kw))
    }
}

#[cfg(test)]
#[path = "safety_test.rs"]
mod safety_test;

